gpt4all-backend/llmodel_c.cpp

```diff
@@ -210,3 +209,57 @@ const char *llmodel_get_implementation_search_path()
 {
     return LLModel::Implementation::implementationsSearchPath().c_str();
 }
+
+struct llmodel_gpu_device* llmodel_available_gpu_devices(llmodel_model model, size_t memoryRequired, int* num_devices)
+{
+    LLModelWrapper *wrapper = reinterpret_cast<LLModelWrapper*>(model);
+    std::vector<LLModel::GPUDevice> devices = wrapper->llModel->availableGPUDevices(memoryRequired);
+
+    // Set the num_devices
+    *num_devices = devices.size();
+
+    if (*num_devices == 0) return nullptr;  // Return nullptr if no devices are found
+
+    // Allocate memory for the output array
+    struct llmodel_gpu_device* output = (struct llmodel_gpu_device*) malloc(*num_devices * sizeof(struct llmodel_gpu_device));
```

Comment:
ideally we should not leak this struct and the name/vendor strings. perhaps associate it with the LLModelWrapper and free it when the LLModelWrapper is freed - you could also stack all the names/vendors into a single buffer (just use a std::string or std::vector) instead of doing a fresh malloc for each